pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -173,52 +209,47 @@ func (w *Watcher) Start() error {
 		return err
 	}
 
-	watcher, err := fsnotify.NewWatcher()
+	fsWatcher, err := fsnotify.NewWatcher()
 	if err != nil {
-		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+		return fmt.Errorf("failed to start plugin fsWatcher, err: %v", err)
 	}
+	w.fsWatcher = fsWatcher
 
-	if err := watcher.Add(w.path); err != nil {
-		watcher.Close()
-		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	if err := fsWatcher.Add(w.path); err != nil {
+		fsWatcher.Close()
+		return fmt.Errorf("failed to watch %s, err: %v", w.path, err)
 	}
 
-	w.watcher = watcher
-
-	if err := w.traversePluginDir(); err != nil {
-		watcher.Close()
+	if err := w.traversePluginDir(w.path); err != nil {
+		fsWatcher.Close()
 		return fmt.Errorf("failed to traverse plugin socket path, err: %v", err)
 	}
 
 	w.wg.Add(1)
```

Comment:
Use of this waitgroup is unclear (I know it's not related to this PR)

Reply:
Do you mind explaining why this is unclear :) ?
This is a standard go patter where you add 1 before starting a goroutine

Reply:
it can be more than 1 so technically it's not a group ;) as a barrier it just needs to be `chan`.

Reply:
I found in `Stop` function, this `wg` is translated to a chan, so it's really not necessary to use a waitgroup.

Reply:
A waitgroup is a standard golang pattern, that is a lot more maintainable than having a channel.

That is, if in the future, this code has to handle a few more more goroutines, a waitgroup allows a simple barrier without having to change a lot of code or basically re-writing a waitgroup using channels.

Which, saying this makes me realize that we should add(1) each time we process an event.

> I found in Stop function, this wg is translated to a chan, so it's really not necessary to use a waitgroup.

That's not the case, the Stop function is written in a way that after waiting 10 seconds on the `Done` call it raises a timeout error. It does not "translate the wg to a channel".

Reply:
@RenaudWasTaken it's typical when you have a *group* of events which you want to wait on. The group concept isn't useful here because `Start()` cannot be called more than once. The translation happens above what you pointed out in a goroutine. the goroutine is there because it needs a `chan` to select as you pointed out.

Reply:
> @RenaudWasTaken it's typical when you have a group of events which you want to wait on. The group concept isn't useful here because Start() cannot be called more than once.

It is useful because it is a lot more maintainable. Just because we only have one cll to `Add` doesn't mean in the future this function won't have more. At that point it's a lot more useful and maintainable to have a waitgroup than to have to either add more `goroutineStopped` channels (i.e: rewriting a waitgroup) or replace the channel mechanism you are suggesting with a waitgroup.

To build on this point I was also suggesting to call `Add` for each event processed. This concrete example should show that it's a lot easier to always work with a waitgroup than to use a `goroutineStopped` channel per goroutine (or whatever pattern using channels).

Independently of maintainibility it's a lot easier to grasp the use of waitgroup than the use of having two or more channels.